base_user_role_company/README.rst

```diff
@@ -0,0 +1,119 @@
+=====================
+User roles by company
+=====================
+
+.. 
+   !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+   !! This file is generated by oca-gen-addon-readme !!
+   !! changes will be overwritten.                   !!
+   !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+   !! source digest: sha256:d53f251191c84826b5c4436215ab4a9ca8699759795f4a2ce11149295f83ef6c
+   !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+
+.. |badge1| image:: https://img.shields.io/badge/maturity-Beta-yellow.png
+    :target: https://odoo-community.org/page/development-status
+    :alt: Beta
+.. |badge2| image:: https://img.shields.io/badge/licence-AGPL--3-blue.png
+    :target: http://www.gnu.org/licenses/agpl-3.0-standalone.html
+    :alt: License: AGPL-3
+.. |badge3| image:: https://img.shields.io/badge/github-OCA%2Fserver--backend-lightgray.png?logo=github
+    :target: https://github.com/OCA/server-backend/tree/16.0/base_user_role_company
+    :alt: OCA/server-backend
+.. |badge4| image:: https://img.shields.io/badge/weblate-Translate%20me-F47D42.png
+    :target: https://translation.odoo-community.org/projects/server-backend-16-0/server-backend-16-0-base_user_role_company
+    :alt: Translate me on Weblate
+.. |badge5| image:: https://img.shields.io/badge/runboat-Try%20me-875A7B.png
+    :target: https://runboat.odoo-community.org/builds?repo=OCA/server-backend&target_branch=16.0
+    :alt: Try me on Runboat
+
+|badge1| |badge2| |badge3| |badge4| |badge5|
+
+Enable User Roles depending on the Companies selected.
+
+A company specific Role will only be enabled
+if it is set for **all** the currently selected companies.
```

Comment:
@dreispt this isn't true since at least v14. To me it would make sense to adapt the code accordingly, or do you as maintainer say better adapt the readme?

Reply:
@hbrunn  agree that the current README does not reflect what the code does.

The current code will poses a threat from a security standpoint. If you are assigned to Sales Manager role in company A only, you can still access to all sales manager functions in company B by just enabling this company together with company A.

I think that it's necessary to adapt the code to the README.

Reply:
@hbrunn  agree that the current README does not reflect what the code does.

The current code will poses a threat from a security standpoint. If you are assigned to Sales Manager role in company A only, you can still access to all sales manager functions in company B by just enabling this company together with company A.

I think that it's necessary to adapt the code to the README.

Reply:
Is there some news about this?

Reply:
fixed this right now

Reply:
@JordiBForgeFlow @sebalix @dreispt but wait, if this is supposed to be a security feature (as opposed to convenience as I saw it up to now), then the module needs to be revamped entirely:

We set the groups only on authentication or when get_session_info is called, that means if I authenticate, select company1 only, I'll have groups specific to company1.

For people just clicking things in the webclient, that's fine. But if I now craft a request for this session with a context setting allowed companies to company1,company2, this request runs with the groups specific to company1, but possibly on data of company2.

As we don't want to update groups for *every* request, and we can't rewrite `base_user_role` to not work by rewriting groups, I think we'll have to save somewhere server side for which companies the last update of groups has taken place, and restrict to/raise if we get allowed companies not in this list? With the current approach you can't be sensibly in company1 with one tab and company2 with another tab and the same user anyways, so this wouldn't really make things worse?

Reply:
With the current data model, I don't see how to fix this issue. Seems as if we'd want to store, a model res.user.group.company that stores for which companies is a user assigned to a group..

Reply:
I think introducing a field `last_cids` on res.users that we write every time we rewrite groups should work. Then we can restrict the incoming allowed companies to the intersection with `last_cids`, and the company switcher then has to call a controller that does updating last_cids + group update in one transaction.

With this we should be able to guarantee you never have groups assigned for companies that shouldn't.

Reply:
Hello here!
I work for an Odoo integrator. I think this module could be quite useful for some clients. I subscribe to the analysis that it doesn't do what it's expected to as soon it is considered as a security module (which is how I see it and how I think it could be useful).
I'm a functional guy so I'm probably missing many technical aspects. Anyways, here is my take about what the module should do : 
- Regarding the Menus / Views / Access Rights, consider the intersection of what the companies ticked in the selector give access to (through the groups associated to the roles and their inherited groups). Example with Access Rights, if you have access to those two groups through companies ticked in the selector : 
![image](https://github.com/OCA/server-backend/assets/106684081/08be2db3-2ba3-4754-9d1e-9d55d067a523)
![image](https://github.com/OCA/server-backend/assets/106684081/b6eda770-16be-4b89-a57f-60a51bbe7f8a)
=> Then you should only get read and write access to sales orders and sale orders lines
- Regarding the record rules : apply an "AND" rule. In the example below we have 2 groups of record rules : 
![image](https://github.com/OCA/server-backend/assets/106684081/7c8f49f3-6f0d-4f66-af4c-ddff666e6ae4)
![image](https://github.com/OCA/server-backend/assets/106684081/d6271c3e-ac13-46e9-a96b-9eeb2fb3a4be)
=> In this example, since we have a record rule restricting the modification to one's own sales and the other to the sales in "draft" state (quotations), if the companies in the selector give access to those two groups, the modification should only be allowed for one's own sales in "draft" state.

What's your opinion ? Happy to discuss this topic with you !

Reply:
what you describe is changing how roles work in the first place. That would involve changing the base module, so very much out of scope here.

Reply:
Hello @hbrunn ,
Thanks for reading my post and coming back to me!
Do you have other solutions in mind ?

Reply:
what I wrote [above](https://github.com/OCA/server-backend/pull/241#discussion_r1495217717), but that also won't happen in this PR I think